### PR TITLE
add --output-logs= option to turbo CLI

### LINKED ISSUE #364: Silence cache hit logs
### Describe the feature you'd like to request

To minimise the output of information in the terminal and make it easier to see what was cached and what wasn't, it'd be useful to be able to minimise unnecessary terminal output. Cache hit logs are often not very useful to be seen in the terminal, as those tasks have successfully passed so you wouldn't usually need to be able to see them.. they're just noise at that point.

### Describe the solution you'd like

A --silenceCachedLogs flag passed to the turborepo cli that allows you to silence the output of logs when there is a cache hit. Ie putting this line in a conditional: https://github.com/vercel/turborepo/blob/788d57506abfe3d1a9d1d9e2e4ac9a2b65998ec4/cli/internal/run/run.go#L457

### Describe alternatives you've considered

A summary table after turborepo has finished outputting which summarises what was cached and what wasn't, and how long each non-cached step took to run. Perhaps that could be done in conjunction with the above?

### PR DESCRIPTION
**Pull request by chelkyl**
* Adds `--output-logs` option with a few mode options to control output logs of tasks run by `turbo run build`
* `full` is default and shows all tasks' process output
* `hash-only` only outputs the hash line which has `cache <miss|hit>, <executing|replaying output> <hash>`
* `new-only` shows all process output for tasks that have not been cached and only the hash line for cached tasks
* `none` hides all tasks' process output

Usage:
`turbo run build --output-logs=new-only`

Without --output-logs or --output-logs=full
image

With --output-logs=hash-only
image

With --output-logs=new-only
image

With --output-logs=none
image

Closes #364

### CODE REVIEW

**Comment by dcherman** [comment]
As a thought, can we also add `turbo logs <hash>` as a way to re-play logs from cached builds?  That way if you do choose to use this flag, you still have a way of inspecting the logs from the completed builds if needed since you'd have the hashes.

**Comment by weyert** [comment]
I am wondering if something like similar to `docker's  `--progress` flag makes more sense here?
> --progress | auto | Set type of progress output (auto, plain, tty). Use plain to show container output

**Comment by jlarmstrongiv** [comment]
Wow, this is exactly what I need 😄 can’t wait to see it merged

I think @dcherman’s idea is an interesting.  All logs are saved in the `node_modules/.cache/turbo/<hash>/.turbo/turbo-<command>.log` if you need to find it in the meantime.

**Comment by chelkyl** [comment]
> I am wondering if something like similar to `docker's `--progress` flag makes more sense here?
> 
> > --progress | auto | Set type of progress output (auto, plain, tty). Use plain to show container output

@weyert That does make a lot of sense! Refactoring.

**Comment by chelkyl** [comment]
> As a thought, can we also add `turbo logs <hash>` as a way to re-play logs from cached builds? That way if you do choose to use this flag, you still have a way of inspecting the logs from the completed builds if needed since you'd have the hashes.

@dcherman Great idea! I'll see what can be done and what other features this command could have.

**Comment by weyert** [comment]
Nice, liking this. How can we progress this PR? @gsoltis

**Comment by gsoltis** [comment]
I like the idea, but I'm a little concerned about the flag name and option values. As I understand it, this is intended to control the behavior of logging when a cache hit occurs. The various options are "replay the logs", "don't replay the logs", and "replay just the first line that shows a cache hit". Does that seem like an accurate summary?

`--progress` on its own doesn't, at least to me, suggest logs behavior. Maybe there's another term we could use? Something like `--cached-logs=[full|silent|name-only]` or something like that? Definitely open to suggestions on this.

**Comment by chelkyl** [comment]
> I like the idea, but I'm a little concerned about the flag name and option values. As I understand it, this is intended to control the behavior of logging when a cache hit occurs. The various options are "replay the logs", "don't replay the logs", and "replay just the first line that shows a cache hit". Does that seem like an accurate summary?

@gsoltis Previously, the `--progress` option only had 2 modes, "replay the logs" and  "replay just the first line that shows a cache hit".

> `--progress` on its own doesn't, at least to me, suggest logs behavior. Maybe there's another term we could use? Something like `--cached-logs=[full|silent|name-only]` or something like that? Definitely open to suggestions on this.

It was based on what docker provides and I wanted to keep it to a single word but I agree that it could be clearer! I have renamed it to `--output-logs` and added more functionality to control both cached and non-cached task output.
See the updated description and let me know what you think!

**Review by gsoltis — CHANGES_REQUESTED** [review]
This is coming along!

Can you also resolve any merge conflicts?

File: cli/internal/run/run.go [hunk 1]
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
 	bail            bool
 	passThroughArgs []string
 	// Restrict execution to only the listed task names. Default false
-	only       bool
-	dryRun     bool
-	dryRunJson bool
+	only bool
+	// Task logs output modes (cached and not cached tasks):
```
[the remark this answers is not in the document]
Comment (chelkyl) [reply cli/internal/run/run.go:444]: Done! Let me know if you prefer other names/values

File: cli/internal/run/run.go [hunk 2]
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
...
+	// full - show all,
+	// hash - only show task hash,
+	// none - show nothing
+	cachedOutputLogsMode    string
```
[the remark this answers is not in the document]
Comment (chelkyl) [reply cli/internal/run/run.go]: Agreed! Done

File: cli/internal/run/run.go [hunk 3]
```diff
@@ -427,9 +432,15 @@ type RunOptions struct {
...
+	notCachedOutputLogsMode string
```
[the remark this answers is not in the document]
Comment (chelkyl) [reply cli/internal/run/run.go]: Agreed! Done

File: cli/internal/run/run.go [hunk 4]
```diff
@@ -559,6 +572,24 @@ func parseRunArgs(args []string, output cli.Ui) (*RunOptions, error) {
 				includDepsSet = true
 			case strings.HasPrefix(arg, "--only"):
 				runOptions.only = true
+			case strings.HasPrefix(arg, "--output-logs"):
```
[the remark this answers is not in the document]
Comment (chelkyl) [reply cli/internal/run/run.go]: Great catch! Fixed

File: cli/internal/run/run.go [hunk 5]
```diff
@@ -559,6 +572,24 @@ func parseRunArgs(args []string, output cli.Ui) (*RunOptions, error) {
...
+				outputLogsMode := arg[len("--output-logs="):]
+				if len(outputLogsMode) > 0 {
```
[the remark this answers is not in the document]
Comment (chelkyl) [reply cli/internal/run/run.go]: That makes sense to me! Done

**Review by gsoltis — APPROVED** [review]

**Comment by chelkyl** [comment]
Thanks!

**Pull request merged**